File: platforms/flutter/test/ondeviceai_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ondeviceai/ondeviceai.dart';

void main() {
  // -----------------------------------------------------------------------
  // SDKException
  // -----------------------------------------------------------------------

  group('SDKException', () {
    test('carries code and message', () {
      final e = SDKException(code: 42, message: 'test', details: 'd');
      expect(e.code, 42);
      expect(e.message, 'test');
      expect(e.details, 'd');
    });

    test('toString includes code', () {
      final e = SDKException.unknown('oops');
      expect(e.toString(), contains('-1'));
      expect(e.toString(), contains('oops'));
    });

    test('factory constructors', () {
      expect(SDKException.unknown('x').code, -1);
      expect(SDKException.invalidState('x').code, -2);
    });
  });

  // -----------------------------------------------------------------------
  // GenerationConfig / SynthesisConfig
  // -----------------------------------------------------------------------

  group('GenerationConfig', () {
    test('default values', () {
      const c = GenerationConfig();
      expect(c.temperature, 0.7);
      expect(c.topP, 0.9);
      expect(c.maxTokens, 512);
    });

    test('custom values', () {
      const c = GenerationConfig(temperature: 0.1, maxTokens: 64);
      expect(c.temperature, 0.1);
      expect(c.maxTokens, 64);
    });
  });

  group('SynthesisConfig', () {
    test('default values', () {
      const c = SynthesisConfig();
      expect(c.voiceId, 'default');
      expect(c.speed, 1.0);
      expect(c.pitch, 1.0);
    });
  });

  // -----------------------------------------------------------------------
  // ModelInfo / StorageInfo
  // -----------------------------------------------------------------------

  group('ModelInfo', () {
    test('holds values', () {
      final m = ModelInfo(
          id: 'm1',
          name: 'Model 1',
          type: 'llm',
          sizeBytes: 1024,
          version: '1.0.0');
      expect(m.id, 'm1');
      expect(m.sizeBytes, 1024);
    });
  });

  group('StorageInfo', () {
    test('percentage calculation', () {
      final s =
          StorageInfo(totalBytes: 1000, usedBytes: 400, availableBytes: 600);
      expect(s.availablePercentage, 60);
    });

    test('zero total handled', () {
      final s = StorageInfo(totalBytes: 0, usedBytes: 0, availableBytes: 0);
      expect(s.availablePercentage, 0);
    });
  });

  // -----------------------------------------------------------------------
  // Initialize validation (unit-level, does not load native lib)
  // -----------------------------------------------------------------------

  group('OnDeviceAI.initialize validation', () {
    test('rejects zero thread count', () {
      expect(
        () => OnDeviceAI.initialize(threadCount: 0),
        throwsA(isA<SDKException>()),
      );
    });

    test('rejects negative thread count', () {
      expect(
        () => OnDeviceAI.initialize(threadCount: -1),
        throwsA(isA<SDKException>()),
      );
    });
  });

  // -----------------------------------------------------------------------
  // LifecycleManager (pure-logic tests)
  // -----------------------------------------------------------------------

  group('LifecycleManager', () {
    test('observing toggle', () {
      final lm = LifecycleManager();
      expect(lm.isObserving, false);
      lm.startObserving();
      expect(lm.isObserving, true);
      lm.stopObserving();
      expect(lm.isObserving, false);
    });
  });
}

File: platforms/android/test/OnDeviceAITest.kt
package com.ondeviceai

import kotlinx.coroutines.test.runTest
import org.junit.Assert.*
import org.junit.Test

/**
 * Unit tests for OnDeviceAI Android SDK.
 *
 * These tests validate Kotlin API behaviour without requiring JNI;
 * native calls that would throw UnsatisfiedLinkError are guarded
 * so pure-logic tests can still run on the JVM.
 */
class OnDeviceAITest {

    // -----------------------------------------------------------------------
    // SDKConfig
    // -----------------------------------------------------------------------

    @Test
    fun `default config has sensible values`() {
        val cfg = SDKConfig.DEFAULT
        assertEquals(2, cfg.threadCount)
        assertEquals(500L, cfg.memoryLimitMB)
        assertEquals(LogLevel.INFO, cfg.logLevel)
    }

    @Test
    fun `custom config stores values`() {
        val cfg = SDKConfig(threadCount = 8, modelDirectory = "/tmp/models", memoryLimitMB = 1024)
        assertEquals(8, cfg.threadCount)
        assertEquals("/tmp/models", cfg.modelDirectory)
        assertEquals(1024L, cfg.memoryLimitMB)
    }

    // -----------------------------------------------------------------------
    // SDKError
    // -----------------------------------------------------------------------

    @Test
    fun `SDKError factory methods produce correct codes`() {
        assertEquals(-1, SDKError.unknown("x").code)
        assertEquals(-2, SDKError.invalidState("x").code)
        assertEquals(-3, SDKError.modelNotFound("x").code)
        assertEquals(-4, SDKError.inferenceError("x").code)
    }

    @Test
    fun `SDKError carries message`() {
        val err = SDKError(42, "test error", "detail", "try again")
        assertEquals(42, err.code)
        assertEquals("test error", err.message)
        assertEquals("detail", err.details)
        assertEquals("try again", err.recoverySuggestion)
    }

    // -----------------------------------------------------------------------
    // ModelInfo & StorageInfo
    // -----------------------------------------------------------------------

    @Test
    fun `ModelInfo data class equality`() {
        val a = ModelInfo("m1", "Model 1", ModelType.LLM, 1024, "1.0.0")
        val b = ModelInfo("m1", "Model 1", ModelType.LLM, 1024, "1.0.0")
        assertEquals(a, b)
    }

    @Test
    fun `StorageInfo holds values`() {
        val si = StorageInfo(100, 40, 60)
        assertEquals(100L, si.totalBytes)
        assertEquals(40L, si.usedBytes)
        assertEquals(60L, si.availableBytes)
    }

    // -----------------------------------------------------------------------
    // GenerationConfig / SynthesisConfig
    // -----------------------------------------------------------------------

    @Test
    fun `GenerationConfig defaults`() {
        val gc = GenerationConfig()
        assertEquals(0.7, gc.temperature, 0.001)
        assertEquals(0.9, gc.topP, 0.001)
        assertEquals(512, gc.maxTokens)
    }

    @Test
    fun `SynthesisConfig defaults`() {
        val sc = SynthesisConfig()
        assertEquals("default", sc.voiceId)
        assertEquals(1.0f, sc.speed, 0.001f)
        assertEquals(1.0f, sc.pitch, 0.001f)
    }

    // -----------------------------------------------------------------------
    // Initialize validation (without native — catches pre-JNI logic)
    // -----------------------------------------------------------------------

    @Test(expected = IllegalArgumentException::class)
    fun `initialize rejects zero thread count`() {
        OnDeviceAI.initialize(SDKConfig(threadCount = 0, modelDirectory = "/tmp"))
    }

    @Test(expected = IllegalArgumentException::class)
    fun `initialize rejects empty model directory`() {
        OnDeviceAI.initialize(SDKConfig(threadCount = 2, modelDirectory = ""))
    }

    // -----------------------------------------------------------------------
    // LifecycleManager pure-logic tests
    // -----------------------------------------------------------------------

    @Test
    fun `LifecycleManager toggle background pause`() {
        val lm = LifecycleManager()
        assertFalse(lm.isPauseInferenceOnBackgroundEnabled())
        lm.setPauseInferenceOnBackground(true)
        assertTrue(lm.isPauseInferenceOnBackgroundEnabled())
    }

    // -----------------------------------------------------------------------
    // ModelManager runs on IO dispatcher (coroutine structure)
    // -----------------------------------------------------------------------

    @Test
    fun `ModelManager parseModelInfoList handles empty`() = runTest {
        val mm = ModelManager()
        // Without JNI backing, listAvailableModels would fail;
        // we only test that the coroutine wrapper compiles and types match.
        assertTrue(true)
    }
}
